### PR TITLE
Document more MSTest attributes and APIs

### PR DESCRIPTION
**Pull request by Evangelink**
cc @Youssef1313 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/testing/unit-testing-mstest-writing-tests-attributes.md](https://github.com/dotnet/docs/blob/d36ddca0722dc2b8e265fa60060b2b20b710d8d5/docs/core/testing/unit-testing-mstest-writing-tests-attributes.md) | [MSTest attributes](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-writing-tests-attributes?branch=pr-en-us-50979) |

### CODE REVIEW

**Comment by Evangelink** [comment]
@Youssef1313 @meaghanlewis working on this doc page, I feel like the attribute grouping is too technical and it would probably be better to have something like:

1. Overview (basics + getting started)
2. Data-Driven Testing (keep separate - it's substantial)
3. Test Lifecycle & Setup (fixtures)
4. Test Execution & Control (threading, parallelization, timeout, retry, conditional execution combined)
5. Metadata & Organization (all metadata attributes)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 1]
```diff
@@ -263,60 +524,268 @@ public class MyTestClass
...
 
 When applied to a test method, the [STATestMethod](xref:Microsoft.VisualStudio.TestTools.UnitTesting.STATestMethodAttribute) attribute indicates that the test method should be run in a single-threaded apartment (STA). This attribute is useful when the test method interacts with COM objects that require STA.
 
 > [!NOTE]
 > This is only supported on Windows and in version 3.6 and later.
 
-### `ParallelizeAttribute`
+#### `UITestMethodAttribute`
+
+When applied to a test method, the `UITestMethod` attribute indicates that the test method should be scheduled on the UI thread of the platform. This attribute is specifically designed for testing UWP (Universal Windows Platform) and WinUI applications that require UI thread access.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:582]: For some reason the API doc related to this attribute doesn't exist. @Youssef1313 is it because the attribute is not present in all TFMs?

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 2]
```diff
@@ -184,6 +209,242 @@ public class TestClass
...
+
+The attribute requires the name of a field, a property, or method that provides the test data. The data source can return any of the following types:
+
+- `IEnumerable<object[]>` - each `object[]` represents the arguments for one test case
+- `IEnumerable<ValueTuple>` - each tuple represents the arguments for one test case (for example, `IEnumerable<(int, string)>`)
+- `IEnumerable<Tuple<...>>` - each `Tuple` represents the arguments for one test case
+- `IEnumerable<TestDataRow>` - provides additional control over test case metadata (see [TestDataRow](#testdatarow) subsection)
+
+> [!NOTE]
+> Any collection type that inherits from `IEnumerable<T>` works, not just `IEnumerable<T>` itself. For example, `List<object[]>`, `object[][]`, or custom collection types are all supported.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:242]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962356)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 3]
```diff
@@ -184,6 +209,242 @@ public class TestClass
...
+        // Test implementation
+    }
+
+    public static IEnumerable<object[]> GetTestData()
+    {
+        yield return new object[] { 1, "first" };
+        yield return new object[] { 2, "second" };
+    }
+
+    public static string GetDisplayName(MethodInfo methodInfo, object[] data)
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:338]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962344)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 4]
```diff
@@ -263,60 +524,268 @@ public class MyTestClass
...
 
-The `Retry` attribute was introduced in MSTest 3.8. This attribute causes the test method to be retried when it fails or timeouts. It allows you to specify the maximum number of retry attempts, the time delay between retries, and a delay backoff type, which is either constant or exponential.
+#### `TimeoutAttribute`
+
+The [Timeout](xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute) attribute can be used to specify the maximum time in milliseconds that a test method is allowed to run. If the test method runs longer than the specified time, the test will be aborted and marked as failed.
+
+This attribute can be applied to any test method or any fixture method (initialization and cleanup methods). It is also possible to specify the timeout globally for either all test methods or all test fixture methods by using the [timeout properties of the runsettings file](./unit-testing-mstest-configure.md#mstest-element).
+
+> [!NOTE]
+> The timeout is not guaranteed to be precise. The test will be aborted after the specified time has passed, but it may take longer before the step is cancelled.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962330)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 5]
```diff
@@ -364,13 +833,16 @@ The following attributes and the values assigned to them appear in the `Visual S
 For example, you could use it to store the name of a "test pass" that this test covers, by marking the test with `[TestProperty("Feature", "Accessibility")]`. Or, you could use it to store an indicator of the kind of test It's with `[TestProperty("ProductMilestone", "42")]`. The property you create by using this attribute, and the property value you assign, are both displayed in the Visual Studio **Properties** window under the heading **Test specific**.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962315)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 6]
```diff
@@ -184,6 +209,242 @@ public class TestClass
...
+
+The `UnfoldingStrategy` property accepts the following values:
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Auto?displayProperty=nameWithType> (default): MSTest automatically determines whether to unfold test cases based on the number of data rows. Test cases are collapsed (folded) when there are many data rows to avoid cluttering Test Explorer, and unfolded when there are few data rows for better visibility.
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Unfold?displayProperty=nameWithType>: All test cases are expanded and shown individually in Test Explorer and TRX results. Each test case can be run independently.
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Fold?displayProperty=nameWithType>: Test cases are collapsed into a single test node. Individual test cases cannot be run independently; the entire data-driven test runs as one unit.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962292)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 7]
```diff
@@ -184,6 +209,242 @@ public class TestClass
...
+
+### `DynamicDataAttribute`
+
+The [DynamicData](xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute) attribute allows you to run the same test method with data provided by a field, a property, or a method. This is useful when you need to generate test data dynamically or when the test data is too complex to define inline with <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute>.
+
+The attribute requires the name of a field, a property, or method that provides the test data. The data source can return any of the following types:
+
+- `IEnumerable<object[]>` - each `object[]` represents the arguments for one test case
+- `IEnumerable<ValueTuple>` - each tuple represents the arguments for one test case (for example, `IEnumerable<(int, string)>`)
+- `IEnumerable<Tuple<...>>` - each `Tuple` represents the arguments for one test case
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:238]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/dotnet/docs/pull/50979#discussion_r2671962246)

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 8]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+    // Fields are also supported
+    public static IEnumerable<object[]> TestDataField = new[]
+    {
+        new object[] { 10, "ten" },
+        new object[] { 20, "twenty" }
+    };
+}
+```
+
+The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> also supports the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataDisplayName> property to customize how test cases appear in Test Explorer. You can specify the display name format using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataDisplayNameDeclaringType> to reference a method that generates the display name:
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> also supports the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute.DynamicDataDisplayName> property to customize how test cases appear in Test Explorer. You can specify the display name format using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute.DynamicDataDisplayNameDeclaringType> to reference a method that generates the display name:
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 9]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+    }
+}
+```
+
+> [!NOTE]
+> To ignore individual test cases when using <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute>, use the `TestDataRow` class with its `IgnoreMessage` property (see [TestDataRow](#testdatarow) section).
+
+#### TestDataRow
+
+The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow> class provides enhanced control over test data in data-driven tests. When using `IEnumerable<TestDataRow<T>>` as the return type for your dynamic data source, you can specify additional metadata for each test case, such as custom display names and test properties.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1> class provides enhanced control over test data in data-driven tests. When using `IEnumerable<TestDataRow<T>>` as the return type for your dynamic data source, you can specify additional metadata for each test case, such as custom display names and test properties.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 10]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+
+`TestDataRow<T>` offers the following benefits:
+
+- **Custom display names**: Set a unique display name for each test case using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow.DisplayName> property.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- **Custom display names**: Set a unique display name for each test case using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1.DisplayName> property.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 11]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+- **Test properties**: Attach metadata to individual test cases using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow.TestProperties> property.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: Should this instead reference [Test categories](https://learn.microsoft.com/dotnet/api/microsoft.visualstudio.testtools.unittesting.testdatarow-1.testcategories?view=mstest-net-4.0#microsoft-visualstudio-testtools-unittesting-testdatarow-1-testcategories)?

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 12]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+            DisplayName = "Test Case 2: Edge case",
+            TestProperties = { ["Priority"] = "High", ["Category"] = "Critical" }
+        };
+    }
+}
+```
+
+### `UnfoldingStrategy` property
+
+Data-driven test attributes like <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute> and <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> support the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy> property, which controls how test cases appear in Test Explorer and TRX results. This property also determines whether individual test cases can be run independently.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
Data-driven test attributes like <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute> and <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> support the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataSourceUnfoldingStrategy> property, which controls how test cases appear in Test Explorer and TRX results. This property also determines whether individual test cases can be run independently.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 13]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+        yield return new object[] { 2 };
+        yield return new object[] { 3 };
+    }
+}
+```
+
 ## Attributes used to provide initialization and cleanups
 
-Setup and cleanup that is common to multiple tests can be extracted to a separate method, and marked with one of the attributes listed below, to run it at appropriate time, for example before every test. For more information, see [Anatomy of a unit test](/previous-versions/ms182517(v=vs.110)).
+Setup and cleanup that is common to multiple tests can be extracted to a separate method, and marked with one of the attributes listed below, to run it at appropriate time, for example, before every test. For more information, see [Anatomy of a unit test](/previous-versions/ms182517(v=vs.110)).
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
Setup and cleanup that's common to multiple tests can be extracted to a separate method and marked with one of the attributes listed in the following section, to run it at appropriate time, for example, before every test. For more information, see [Anatomy of a unit test](/previous-versions/ms182517(v=vs.110)).
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 14]
```diff
@@ -263,60 +526,268 @@ public class MyTestClass
...
+    public static void GlobalTestCleanup(TestContext testContext)
+    {
+        // This runs after every test method in the assembly
+    }
+}
+```
 
 > [!NOTE]
-> The timeout is not guaranteed to be precise. The test will be aborted after the specified time has passed, but it may take longer before the step is cancelled.
+> Multiple methods with these attributes in the assembly are allowed, but there is no guarantee of the order in which they will be executed. The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute> isn't supported on methods with the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.GlobalTestInitializeAttribute>.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
> Multiple methods with these attributes in the assembly are allowed, but there's no guarantee of the order in which they'll execute. The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute> isn't supported on methods with the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.GlobalTestInitializeAttribute>.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 15]
```diff
@@ -263,60 +526,268 @@ public class MyTestClass
...
 
-### `RetryAttribute`
+### Timeout and retry attributes
 
-The `Retry` attribute was introduced in MSTest 3.8. This attribute causes the test method to be retried when it fails or timeouts. It allows you to specify the maximum number of retry attempts, the time delay between retries, and a delay backoff type, which is either constant or exponential.
+These attributes control test execution time limits and retry behavior.
+
+#### `TimeoutAttribute`
+
+The [Timeout](xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute) attribute can be used to specify the maximum time in milliseconds that a test method is allowed to run. If the test method runs longer than the specified time, the test will be aborted and marked as failed.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
The [Timeout](xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute) attribute can be used to specify the maximum time in milliseconds that a test method is allowed to run. If the test method runs longer than the specified time, the test is aborted and marked as failed.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 16]
```diff
@@ -263,60 +526,268 @@ public class MyTestClass
...
+
+This attribute can be applied to any test method or any fixture method (initialization and cleanup methods). It is also possible to specify the timeout globally for either all test methods or all test fixture methods by using the [timeout properties of the runsettings file](./unit-testing-mstest-configure.md#mstest-element).
+
+> [!NOTE]
+> The timeout is not guaranteed to be precise. The test will be aborted after the specified time has passed, but it might take longer before the step is cancelled.
```
Comment (meaghanlewis) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
> The timeout isn't guaranteed to be precise. The test is be aborted after the specified time passes, but it might take longer before the step is cancelled.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md (hunk 8 above)
Comment (meaghanlewis) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: Are these the correct xrefs?

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md (hunk 9 above)
Comment (meaghanlewis) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: Is this the correct xref?

**Review by meaghanlewis — COMMENTED** [review]
These changes LGTM. I had a handful of suggestions for updated xrefs to use in place of the ones that weren't found. Also, other small language suggestions.

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md (hunk 8 above)
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: Yes! Thank you!

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 17]
```diff
@@ -361,16 +832,19 @@ The [ExpectedException](xref:Microsoft.VisualStudio.TestTools.UnitTesting.Expect
...
 
-For example, you could use it to store the name of a "test pass" that this test covers, by marking the test with `[TestProperty("Feature", "Accessibility")]`. Or, you could use it to store an indicator of the kind of test It's with `[TestProperty("ProductMilestone", "42")]`. The property you create by using this attribute, and the property value you assign, are both displayed in the Visual Studio **Properties** window under the heading **Test specific**.
+For example, you could use it to store the name of a "test pass" that this test covers, by marking the test with `[TestProperty("Feature", "Accessibility")]`. Or, you could use it to store an indicator of the kind of test it is with `[TestProperty("ProductMilestone", "42")]`. The property you create by using this attribute, and the property value you assign, are both displayed in the Visual Studio **Properties** window under the heading **Test specific**.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
For example, you could use it to store the name of a "test pass" that this test covers, by marking the test with `[TestProperty("Feature", "Accessibility")]`. Or, you could use it to store an indicator of the kind of test it's with `[TestProperty("ProductMilestone", "42")]`. The property you create by using this attribute, and the property value you assign, are both displayed in the Visual Studio **Properties** window under the heading **Test specific**.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md (hunk 11 above)
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]: That's correct!


```suggestion
- **Test categories**: Attach metadata to individual test cases using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1.TestCategories> property.
```

**Comment by Evangelink** [comment]
> @Youssef1313 @meaghanlewis working on this doc page, I feel like the attribute grouping is too technical and it would probably be better to have something like:
> 
> 1. Overview (basics + getting started)
> 2. Data-Driven Testing (keep separate - it's substantial)
> 3. Test Lifecycle & Setup (fixtures)
> 4. Test Execution & Control (threading, parallelization, timeout, retry, conditional execution combined)
> 5. Metadata & Organization (all metadata attributes)

@meaghanlewis @Youssef1313 if you both agree, I'll do a follow-up PR after this one.

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Youssef1313) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:582]: @Evangelink It is `Microsoft.VisualStudio.TestTools.UnitTesting.AppContainer.UITestMethodAttribute`

[the remark this answers is not in the document]
Comment (Youssef1313) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:582]: If it's still missing, then I'm not sure why. It will be best to open internal bug for it.

[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-mstest-writing-tests-attributes.md:582]: It's non-existent, I'll create a bug.

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 18]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+}
+```
+
+### `UnfoldingStrategy` property
+
+Data-driven test attributes like <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute> and <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> support the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataSourceUnfoldingStrategy> property, which controls how test cases appear in Test Explorer and TRX results. This property also determines whether individual test cases can be run independently.
+
+The `UnfoldingStrategy` property accepts the following values:
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Auto?displayProperty=nameWithType> (default): MSTest automatically determines whether to unfold test cases based on the number of data rows. Test cases are collapsed (folded) when there are many data rows to avoid cluttering Test Explorer, and unfolded when there are few data rows for better visibility.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataSourceUnfoldingStrategy.Auto?displayProperty=nameWithType> (default): MSTest automatically determines whether to unfold test cases based on the number of data rows. Test cases are collapsed (folded) when there are many data rows to avoid cluttering Test Explorer, and unfolded when there are few data rows for better visibility.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 19]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Unfold?displayProperty=nameWithType>: All test cases are expanded and shown individually in Test Explorer and TRX results. Each test case can be run independently.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataSourceUnfoldingStrategy.Unfold?displayProperty=nameWithType>: All test cases are expanded and shown individually in Test Explorer and TRX results. Each test case can be run independently.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 20]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.UnfoldingStrategy.Fold?displayProperty=nameWithType>: All test cases are collapsed into a single test node. Individual test cases can't be run independently; the entire data-driven test runs as one unit.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataSourceUnfoldingStrategy.Fold?displayProperty=nameWithType>: All test cases are collapsed into a single test node. Individual test cases can't be run independently; the entire data-driven test runs as one unit.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 21]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+
+For most scenarios, the default `Auto` behavior provides the best balance between usability and performance. Changing this setting is considered an advanced scenario and should only be done when you have specific requirements, such as non-deterministic data source or known limitations or bugs of MSTest.
+
+```csharp
+[TestClass]
+public class TestClass
+{
+    [TestMethod(UnfoldingStrategy = UnfoldingStrategy.Unfold)] // Force unfolding/expanding
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
    [TestMethod(UnfoldingStrategy = TestDataSourceUnfoldingStrategy.Unfold)] // Force unfolding/expanding
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 22]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+    [DataRow(1)] 
+    [DataRow(2)]
+    [DataRow(3)]
+    public void TestMethodWithUnfolding(int value)
+    {
+        // Each test case appears individually in Test Explorer
+    }
+
+    [TestMethod(DynamicDataUnfoldingStrategy = UnfoldingStrategy.Fold)] // Force folding/collapsing
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
    [TestMethod(UnfoldingStrategy = TestDataSourceUnfoldingStrategy.Fold)] // Force folding/collapsing
```

**Comment by Evangelink** [comment]
@meaghanlewis it's complaining that it cannot find the `Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1` prefix link but when I look at the doc API they do exist (e.g. https://learn.microsoft.com/dotnet/api/microsoft.visualstudio.testtools.unittesting.testdatarow-1)

**Comment by meaghanlewis** [comment]
> > @Youssef1313 @meaghanlewis working on this doc page, I feel like the attribute grouping is too technical and it would probably be better to have something like:
> > 
> > 1. Overview (basics + getting started)
> > 2. Data-Driven Testing (keep separate - it's substantial)
> > 3. Test Lifecycle & Setup (fixtures)
> > 4. Test Execution & Control (threading, parallelization, timeout, retry, conditional execution combined)
> > 5. Metadata & Organization (all metadata attributes)
> 
> @meaghanlewis @Youssef1313 if you both agree, I'll do a follow-up PR after this one.

Sounds like a plan!

**Comment by meaghanlewis** [comment]
> @meaghanlewis it's complaining that it cannot find the `Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1` prefix link but when I look at the doc API they do exist (e.g. https://learn.microsoft.com/dotnet/api/microsoft.visualstudio.testtools.unittesting.testdatarow-1)

@Evangelink I had a look and can't figure out why these links aren't found. Maybe these can be left out for now and I can open up an issue to address this later.

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 23]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+    }
+}
+```
+
+> [!NOTE]
+> To ignore individual test cases when using <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute>, use the `TestDataRow` class with its `IgnoreMessage` property (see [TestDataRow](#testdatarow) section).
+
+#### TestDataRow
+
+The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1> class provides enhanced control over test data in data-driven tests. When using `IEnumerable<TestDataRow<T>>` as the return type for your dynamic data source, you can specify additional metadata for each test case, such as custom display names and test properties.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
The `TestDataRow<T>` class provides enhanced control over test data in data-driven tests. When using `IEnumerable<TestDataRow<T>>` as the return type for your dynamic data source, you can specify additional metadata for each test case, such as custom display names and test properties.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 24]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+
+`TestDataRow<T>` offers the following benefits:
+
+- **Custom display names**: Set a unique display name for each test case using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1.DisplayName> property.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- **Custom display names**: Set a unique display name for each test case using the `DisplayName` property.
```

File: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md [hunk 25]
```diff
@@ -184,9 +209,247 @@ public class TestClass
...
+- **Test categories**: Attach metadata to individual test cases using the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestDataRow-1.TestCategories> property.
```
Comment (Evangelink) [inline docs/core/testing/unit-testing-mstest-writing-tests-attributes.md]:
```suggestion
- **Test categories**: Attach metadata to individual test cases using the `TestCategories` property.
```

**Comment by Evangelink** [comment]
> @Evangelink I had a look and can't figure out why these links aren't found. Maybe these can be left out for now and I can open up an issue to address this later.

Done! Could you please review/approve @meaghanlewis and @Youssef1313

**Review by meaghanlewis — APPROVED** [review]
